### PR TITLE
Add a database check in the new app generator

### LINKED ISSUE #291: New app with --database check type
Why we need duplication for pg and sqlite in https://github.com/lotus/lotus/blob/master/lib/lotus/generators/application/container.rb#L154-L157 
```
 when 'postgresql', 'postgres'
    "postgres://localhost/#{app_name}"
 when 'sqlite', 'sqlite3'
   "sqlite://db/#{Shellwords.escape(app_name)}"
```

Maybe we should use `postgresql`, `pg` and `sqlite3` instead?
and what about specify `file` type too and generate exception if user tries generate with wrong type.

```
       def database_base_uri
          case @database
          when 'mysql'
            "mysql://localhost/#{app_name}"
          when 'mysql2'
            "mysql2://localhost/#{app_name}"
          when 'postgresql', 'postgres'
            "postgres://localhost/#{app_name}"
          when 'sqlite', 'sqlite3'
            "sqlite://db/#{Shellwords.escape(app_name)}"
          when 'memory'
            "memory://localhost/#{app_name}"
          when 'file'
            "file:///db/#{app_name}"
          else
              railse "#{@database} type is not found"
          end
        end
```
because it's so strange when app is generated with wrong db option for example
`lotus new app --database=postgr` I've tried it and have been confused

.gitignore should not has `db` if is not file database but it has
```
/db/favorites_development
/db/favorites_test
```

### PR DESCRIPTION
**Pull request by mvader**
I'm not sure about the ```autorun``` part. If someone has a better idea for that part it would be welcomed.

Closes #291

### CODE REVIEW

**Comment by jodosha** [comment]
@mvader Thank you for this PR. What's the intent of `autorun`?

**Comment by mvader** [comment]
@jodosha Stop the ```command.start``` run in the ```before```. It throws an exception there and it does not get to the ```it```

-- 
Miguel Molina
Sent with Airmail

On 9 Jul 2015 at 15:23:08, Luca Guidi (notifications@github.com) wrote:

@mvader Thank you for this PR. What's the intent of autorun?

**Comment by mli-max** [comment]
@mvader Can you fix also `.gitignore` as I mentioned in #291?
It should be blank if we don't use filesystem type

**Comment by mli-max** [comment]
@mvader I think you should create test out of block where `capture_io` is called

**Comment by jodosha** [comment]
@mvader I see. `before` blocks are designed to DRY common code. With your new test, it's no longer the case. Please remove `autorun` and inline that call in each `it` block. Simple design always wins. ;)

**Comment by mvader** [comment]
@jodosha that seems the best option. I will update now that and ti .gitignore thing that Maksim said.

-- 
Miguel Molina
Sent with Airmail

On 9 Jul 2015 at 17:45:17, Luca Guidi (notifications@github.com) wrote:

@mvader I see. before blocks are designed to DRY common code. With your new test, it's no longer the case. Please remove autorun and inline that call in each it block. Simple design always wins. ;)

**Comment by mvader** [comment]
If I run the tests multiple times with this change they randomly fail with the following two issues:

```
  1) Failure:
A full stack Lotus application::RESTful CRUD#test_0003_handles edit action [/Users/mvader/lotus/test/integration/full_stack_test.rb:188]:
Expected "<!DOCTYPE html>\n<html>\n  <head>\n    <title>Collaboration</title>\n  </head>\n  <body>\n    <form action=\"/books/1\" method=\"POST\" accept-charset=\"utf-8\" id=\"book-form\">\n<input type=\"hidden\" name=\"_method\" value=\"PATCH\">\n<input type=\"text\" name=\"book[name]\" id=\"book-name\" value=\"The path to success\" placeholder=\"Name\">\n<button type=\"submit\">Update</button>\n</form>\n\n  </body>\n</html>\n" to include "<input type=\"hidden\" name=\"_csrf_token\" value=\"t0k3n\">".


  2) Failure:
A full stack Lotus application::RESTful CRUD#test_0004_handles new action [/Users/mvader/lotus/test/integration/full_stack_test.rb:198]:
Expected "<!DOCTYPE html>\n<html>\n  <head>\n    <title>Collaboration</title>\n  </head>\n  <body>\n    <form action=\"/books\" method=\"POST\" accept-charset=\"utf-8\" id=\"book-form\">\n<input type=\"text\" name=\"book[name]\" id=\"book-name\" value=\"\" placeholder=\"Name\">\n<button type=\"submit\">Create</button>\n</form>\n\n  </body>\n</html>\n" to include "<input type=\"hidden\" name=\"_csrf_token\" value=\"t0k3n\">".


  3) Error:
A full stack Lotus application::CSRF Protection#test_0003_handles destroy action:
NameError: uninitialized constant Collaboration::Views::Authors
    /usr/local/lib/ruby/gems/2.2.0/bundler/gems/utils-0bda3b4a5b6b/lib/lotus/utils/class.rb:49:in `const_get'
    /usr/local/lib/ruby/gems/2.2.0/bundler/gems/utils-0bda3b4a5b6b/lib/lotus/utils/class.rb:49:in `load!'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:71:in `view_for'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:44:in `_render_action'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:37:in `_render'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:28:in `render'
    /Users/mvader/lotus/lib/lotus/application.rb:216:in `call'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/test.rb:94:in `delete'
    /Users/mvader/lotus/test/integration/full_stack_test.rb:256:in `block (3 levels) in <top (required)>'


  4) Error:
A full stack Lotus application::CSRF Protection#test_0002_handles update action:
NameError: uninitialized constant Collaboration::Views::Authors
    /usr/local/lib/ruby/gems/2.2.0/bundler/gems/utils-0bda3b4a5b6b/lib/lotus/utils/class.rb:49:in `const_get'
    /usr/local/lib/ruby/gems/2.2.0/bundler/gems/utils-0bda3b4a5b6b/lib/lotus/utils/class.rb:49:in `load!'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:71:in `view_for'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:44:in `_render_action'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:37:in `_render'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:28:in `render'
    /Users/mvader/lotus/lib/lotus/application.rb:216:in `call'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/test.rb:85:in `patch'
    /Users/mvader/lotus/test/integration/full_stack_test.rb:249:in `block (3 levels) in <top (required)>'


  5) Error:
A full stack Lotus application::CSRF Protection#test_0001_handles create action:
NameError: uninitialized constant Collaboration::Views::Authors
    /usr/local/lib/ruby/gems/2.2.0/bundler/gems/utils-0bda3b4a5b6b/lib/lotus/utils/class.rb:49:in `const_get'
    /usr/local/lib/ruby/gems/2.2.0/bundler/gems/utils-0bda3b4a5b6b/lib/lotus/utils/class.rb:49:in `load!'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:71:in `view_for'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:44:in `_render_action'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:37:in `_render'
    /Users/mvader/lotus/lib/lotus/rendering_policy.rb:28:in `render'
    /Users/mvader/lotus/lib/lotus/application.rb:216:in `call'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
    /usr/local/lib/ruby/gems/2.2.0/gems/rack-test-0.6.3/lib/rack/test.rb:67:in `post'
    /Users/mvader/lotus/test/integration/full_stack_test.rb:242:in `block (3 levels) in <top (required)>'
```

And:

```
/usr/local/lib/ruby/gems/2.2.0/bundler/gems/model-0f4d0619609f/lib/lotus/model/config/adapter.rb:101:in `rescue in instantiate_adapter': Cannot instantiate adapter of Lotus::Model::Adapters::SqlAdapter (Sequel::Database.connect takes either a Hash or a String, given: nil) (RuntimeError)
	from /usr/local/lib/ruby/gems/2.2.0/bundler/gems/model-0f4d0619609f/lib/lotus/model/config/adapter.rb:95:in `instantiate_adapter'
	from /usr/local/lib/ruby/gems/2.2.0/bundler/gems/model-0f4d0619609f/lib/lotus/model/config/adapter.rb:81:in `build'
	from /usr/local/lib/ruby/gems/2.2.0/bundler/gems/model-0f4d0619609f/lib/lotus/model/configuration.rb:256:in `_build_adapter'
	from /usr/local/lib/ruby/gems/2.2.0/bundler/gems/model-0f4d0619609f/lib/lotus/model/configuration.rb:76:in `load!'
	from /usr/local/lib/ruby/gems/2.2.0/bundler/gems/model-0f4d0619609f/lib/lotus/model.rb:87:in `load!'
	from /Users/mvader/lotus/tmp/integration/cli/generate/delivery/lib/delivery.rb:46:in `<top (required)>'
	from /Users/mvader/lotus/tmp/integration/cli/generate/delivery/config/environment.rb:4:in `require_relative'
	from /Users/mvader/lotus/tmp/integration/cli/generate/delivery/config/environment.rb:4:in `<top (required)>'
	from /Users/mvader/lotus/lib/lotus/environment.rb:342:in `require'
	from /Users/mvader/lotus/lib/lotus/environment.rb:342:in `require_application_environment'
	from /Users/mvader/lotus/lib/lotus/generators/migration.rb:31:in `initialize'
	from /Users/mvader/lotus/lib/lotus/commands/generate.rb:79:in `new'
	from /Users/mvader/lotus/lib/lotus/commands/generate.rb:79:in `generator'
	from /Users/mvader/lotus/lib/lotus/commands/generate.rb:44:in `start'
	from /Users/mvader/lotus/lib/lotus/cli.rb:96:in `generate'
	from /usr/local/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /usr/local/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /usr/local/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /usr/local/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /Users/mvader/lotus/bin/lotus:4:in `<top (required)>'
	from /usr/local/lib/ruby/gems/2.2.0/bin/lotus:23:in `load'
	from /usr/local/lib/ruby/gems/2.2.0/bin/lotus:23:in `<main>'
```

However, if I create a new valid app after the test that raises an exception everything is fine.
```ruby
Lotus::Commands::New.new('chirp', Lotus::Environment.new(Hash[architecture: architecture, application: 'web', application_base_url: '/', lotus_head: false, database: 'filesystem']), Lotus::Cli.new)
```

Is there anything that depends on the previous app created in a test? It's not the changes on my code that broke that because it happens in two completely different tests that I have not modified.

**Comment by mvader** [comment]
Nevermind, fixed it. @mli-max also finished the .gitignore thing.

File: lib/lotus/generators/application/app.rb [hunk 1]
```diff
@@ -110,7 +110,8 @@ def start
           end
 
           unless git_dir_present?
-            cli.template(source.join('gitignore.tt'), target.join('.gitignore'), opts)
+            src_gitignore = if database_type == :file_system then 'gitignore.tt' else '.gitignore' end
```
Comment (joneslee85) [inline lib/lotus/generators/application/app.rb]: What do you think about changing to a shorter form `database_type == :file_system ? 'blah' : 'ma'`?

**Comment by mvader** [comment]
@joneslee85 done!

**Comment by joneslee85** [comment]
@mvader :+1:

**Comment by jodosha** [comment]
@mvader Thank you, merging this. :+1:

**Comment by mli-max** [comment]
Great!

**Pull request merged**